src/syscall/syscall_libc.go

```diff
@@ -235,20 +235,41 @@ func Mprotect(b []byte, prot int) (err error) {
 }
 
 func Environ() []string {
-	environ := libc_environ
-	var envs []string
-	for *environ != nil {
-		// Convert the C string to a Go string.
-		length := libc_strlen(*environ)
+	// calculate total memory required
+	var length uintptr
+	var vars int
+	for environ := libc_environ; *environ != nil; {
+		length += libc_strlen(*environ)
+		vars++
+		environ = (*unsafe.Pointer)(unsafe.Pointer(uintptr(unsafe.Pointer(environ)) + unsafe.Sizeof(environ)))
+	}
+
+	// allocate our backing slice for the strings
+	b := make([]byte, length)
+	// and the slice we're going to return
+	envs := make([]string, 0, vars)
+
+	// loop over the environment again, this time copying over the data to the backing slice
+	for environ := libc_environ; *environ != nil; {
+		length = libc_strlen(*environ)
+		// construct a Go string pointing at the libc-allocated environment variable data
 		var envVar string
 		rawEnvVar := (*struct {
 			ptr    unsafe.Pointer
 			length uintptr
 		})(unsafe.Pointer(&envVar))
 		rawEnvVar.ptr = *environ
 		rawEnvVar.length = length
-		envs = append(envs, envVar)
-		// This is the Go equivalent of "environ++" in C.
+		// pull off the number of bytes we need for this environment variable
+		var bs []byte
+		bs, b = b[:length], b[length:]
+		// copy over the bytes to the Go heap
+		copy(bs, envVar)
+		// convert trimmed slice to string
+		s := *(*string)(unsafe.Pointer(&bs))
```

Comment:
This is a dirty trick... but I have to admit, this whole function isn't exactly clean and I wrote it originally.
(Dirty because it's not guaranteed that slice/string layout will stay the way it is now - although I don't see why it would need to change).